File: backend/fix.py
import os

path = r"c:\Users\Muneesha\Desktop\build-track\Build-Track\backend\routes\transactionRoutes.js"
with open(path, "r", encoding="utf-8") as f:
    text = f.read()

text = text.replace("if (results.successes.length > 0) {", "if (results.successCount > 0) {")

backup_routes = """
// --- Added backup routes ---
router.post("/backup-csv", requirePermission(["manage_expenses", "add_entries"]), async (req, res) => {
  try {
    const transactions = await Transaction.find({ createdBy: req.user._id }).lean();
    const fs = require('fs');
    const path = require('path');
    const backupPath = path.join(__dirname, '..', 'backups');
    if (!fs.existsSync(backupPath)) {
      fs.mkdirSync(backupPath);
    }
    const userBackupFile = path.join(backupPath, `backup_${req.user._id}.json`);
    fs.writeFileSync(userBackupFile, JSON.stringify(transactions));
    res.json({ message: "Backup successful" });
  } catch (error) {
    console.error("Backup error:", error);
    res.status(500).json({ message: "Backup failed" });
  }
});

router.post("/revert-csv", requirePermission(["manage_expenses", "add_entries"]), async (req, res) => {
  try {
    const fs = require('fs');
    const path = require('path');
    const userBackupFile = path.join(__dirname, '..', 'backups', `backup_${req.user._id}.json`);
    if (!fs.existsSync(userBackupFile)) {
      return res.status(404).json({ message: "No backup found to revert" });
    }
    const backupData = JSON.parse(fs.readFileSync(userBackupFile, 'utf8'));
    
    const session = await mongoose.startSession();
    session.startTransaction();
    try {
      await Transaction.deleteMany({ createdBy: req.user._id }).session(session);
      if (backupData.length > 0) {
        await Transaction.insertMany(backupData, { session });
      }
      await session.commitTransaction();
      res.json({ message: "Revert successful" });
    } catch (e) {
      await session.abortTransaction();
      throw e;
    } finally {
      session.endSession();
    }
  } catch (error) {
    console.error("Revert error:", error);
    res.status(500).json({ message: "Revert failed" });
  }
});
// -----------------------------

module.exports = router;
"""

text = text.replace("module.exports = router;", backup_routes)

with open(path, "w", encoding="utf-8") as f:
    f.write(text)
print("Done")

File: backend/fix.js
const fs = require('fs');
let code = fs.readFileSync('routes/transactionRoutes.js', 'utf8');
const startIndex = code.indexOf('router.get("/:id"');
if (startIndex !== -1) {
    const blockEnd = code.indexOf('router.post("/"', startIndex);
    let block = code.substring(startIndex, blockEnd);
    block = block.replace(/{ await session\.abortTransaction\(\); return res\.status\(403\)/g, 'return res.status(403)');
    block = block.replace(/} \}/g, '}'); // wait, the original was: { await ...; return ... }
    
    // Actually just replace exactly:
    block = block.split("{ await session.abortTransaction(); return res.status(403).json({ message: 'Access denied to this transaction' }); }").join("return res.status(403).json({ message: 'Access denied to this transaction' });");

    code = code.substring(0, startIndex) + block + code.substring(blockEnd);
    fs.writeFileSync('routes/transactionRoutes.js', code);
}
